### PR TITLE
Add `--debug` and cache downloads better

### LINKED ISSUE #237: Add `--debug` to `build_tarballs.jl` to drop into a shell
It shouldn't be too hard to add a `--debug` option to `build_tarballs.jl` such that we can drop into a `runshell()` environment within a platform that just failed.

### PR DESCRIPTION
**Pull request by staticfloat**
Closes #237

### CODE REVIEW

File: deps/sandbox.c [hunk 1]
```diff
@@ -480,6 +500,8 @@ static int sandbox_main(const char * root_dir, const char * new_cd, int sandbox_
         // If it was the main pid that exited, return as well.
         return WIFEXITED(status) ? WEXITSTATUS(status) : 1;
       }
+      // Sleep for 100ms so that we're not chewing through CPU cycles
+      usleep(100000);
```
Comment (ararslan) [inline deps/sandbox.c]: :zzz:

**Comment by vchuravy** [comment]
I can confirm that fde03a8 does what it supposed to do.

File: deps/sandbox.c (hunk 1 above)
Comment (Keno) [reply deps/sandbox.c]: We should just stop passing WNOHANG to waitpid

Comment (staticfloat) [reply deps/sandbox.c]: Yeah, duh.  For some reason I thought we were doing more work within this `while` loop.....

**Comment by staticfloat** [comment]
Quick summary of changes here:

* `--debug` allows entering an interactive session when a `build_tarballs.jl` run fails.  The interactive environment does its best to pretend it's at the last step that was executed within the buildscript (e.g. at the point of script exit, it saves all environment variables, mimics a bash history and saves the current working directory) so that when you get dropped into a shell, you can just `<up arrow> <enter>` and run the command that failed.

* The above is possible because we now mount another workspace at `/meta`, which holds things like `.bash_history` and `.env`, the file that gets sourced by `bash` when the debug shell is opened.

* `autobuild()` now takes in a `downloads_dir` kwarg, that allows `build_tarballs.jl` to download dependencies to a shared directory, which should speed up building multiple platforms at once, as the dependencies will be downloaded only once now, instead of once per platform.  This relies on https://github.com/JuliaPackaging/BinaryProvider.jl/commit/4ec2ce0fcc9f7f9cbc5c149f795f74299cefccda, so we'll have to tag a new `BinaryProvider` release before we can tag a release here.  I'll probably do that after I fix `BinaryProvider` tests on 0.7.

**Pull request merged**